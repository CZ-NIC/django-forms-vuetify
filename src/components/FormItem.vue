<template>
    <div>
        <v-textarea
            v-if="field.widget.name === 'textarea'"
            v-bind="getCommonProperties()"
        />

        <template v-else-if="field.widget.name === 'input'">
            <v-text-field
                v-if="field.widget.type === 'text'"
                v-bind="getCommonProperties()"
            />

            <v-text-field
                v-else-if="field.widget.type === 'password'"
                :type="show_password ? 'text' : 'password'"
                :append-icon="show_password ? 'visibility' : 'visibility_off'"
                v-bind="getCommonProperties()"
                @click:append="show_password = !show_password"
            />

            <v-checkbox
                v-else-if="field.widget.type === 'checkbox'"
                v-bind="getCommonProperties()"
            />
            <div v-else>
                Input item of {{ field.widget.type }} type is not supported yet.
                {{ printError(field.widget) }}
            </div>
        </template>

        <div v-else>
            Item with widget: {{ `name=${field.widget.name}, type=${field.widget.type}` }} is not supported yet.
            {{ printError(field.widget) }}
        </div>
    </div>
</template>

<script lang="ts">
import 'reflect-metadata'
import { Vue, Component, Prop } from 'vue-property-decorator'

export interface WidgetInterface {
    name: string;
    type?: string;
}

export type FieldValue = string | boolean | null | undefined

export interface FieldInterface {
    name: string;
    label: string;
    help_text?: string;
    value?: FieldValue;
    required?: boolean;
    errors?: string[];
    widget: WidgetInterface;
}

@Component
export default class FormItem extends Vue {
    @Prop({default: {widget: {}} })
    readonly field!: FieldInterface

    private show_password = false

    private getCommonProperties() {
        return {
            name: this.field.name,
            label: this.field.label,
            'error-messages': this.field.errors || [],
            hint: this.field.help_text,
            rules: [(val: string) => this.field.required && val !== ''],
            value: this.field.value,
        }
    }

    private printError(widget: WidgetInterface): void {
        console.error(`Item with widget: { name=${widget.name}, type=${widget.type} } is not supported yet.`)
    }
}
</script>
