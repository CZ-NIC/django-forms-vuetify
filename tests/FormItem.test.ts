import { Wrapper, shallowMount } from '@vue/test-utils'
import FormItem from '@/components/FormItem.vue'

describe('FormItem snapshot', () => {

    function createWrapperForField(field: object): Wrapper<FormItem> {
        return shallowMount(FormItem, {
            propsData: {
                field
            },
        })
    }

    test('matches snapshot - text input', () => {
        const wrapper: Wrapper<FormItem> = createWrapperForField({
            name: 'test_input',
            label: 'Test label',
            help_text: 'Help for test input',
            value: 'test value',
            required: true,
            errors: ['Error'],
            widget: {name: 'input', type: 'text'},
        })
        expect(wrapper).toMatchSnapshot()
    })

    test('matches snapshot - checkbox', () => {
        const wrapper: Wrapper<FormItem> = createWrapperForField({
            name: 'test_checkbox',
            label: 'Test label',
            help_text: 'Help for test checkbox',
            value: 'test value',
            required: true,
            errors: ['Error'],
            widget: {name: 'input', type: 'checkbox'},
        })
        expect(wrapper).toMatchSnapshot()
    })

    test('matches snapshot - textarea', () => {
        const wrapper: Wrapper<FormItem> = createWrapperForField({
            name: 'test_textarea',
            label: 'Test label',
            help_text: 'Help for test textarea',
            value: 'test value',
            required: true,
            errors: ['Error'],
            widget: {name: 'textarea'},
        })
        expect(wrapper).toMatchSnapshot()
    })
})
