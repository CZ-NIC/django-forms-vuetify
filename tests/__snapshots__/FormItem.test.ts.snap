// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormItem snapshot matches snapshot - checkbox 1`] = `
<div>
  <v-checkbox-stub errorcount="1" errormessages="Error" messages="" rules="val => this.field.required &amp;&amp; val !== ''" successmessages="" value="test value" backgroundcolor="" hint="Help for test checkbox" label="Test label" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn" name="test_checkbox"></v-checkbox-stub>
</div>
`;

exports[`FormItem snapshot matches snapshot - password input 1`] = `
<div>
  <v-text-field-stub errorcount="1" errormessages="Error" messages="" rules="val => this.field.required &amp;&amp; val !== ''" successmessages="" value="test value" appendicon="visibility_off" backgroundcolor="" hint="Help for test password input" label="Test label" loaderheight="2" clearicon="$clear" type="password" name="test_password_input"></v-text-field-stub>
</div>
`;

exports[`FormItem snapshot matches snapshot - text input 1`] = `
<div>
  <v-text-field-stub errorcount="1" errormessages="Error" messages="" rules="val => this.field.required &amp;&amp; val !== ''" successmessages="" value="test value" backgroundcolor="" hint="Help for test input" label="Test label" loaderheight="2" clearicon="$clear" type="text" name="test_input"></v-text-field-stub>
</div>
`;

exports[`FormItem snapshot matches snapshot - textarea 1`] = `
<div>
  <v-textarea-stub errorcount="1" errormessages="Error" messages="" rules="val => this.field.required &amp;&amp; val !== ''" successmessages="" value="test value" backgroundcolor="" hint="Help for test textarea" label="Test label" loaderheight="2" clearicon="$clear" type="text" rowheight="24" rows="5" name="test_textarea"></v-textarea-stub>
</div>
`;
